### PR TITLE
Bug:107963 - (ZCS-1461) Fixed Host header injection vulnerability [CWE-20] in Zimbra's Proxy configuration

### PR DESCRIPTION
**Pull request by shri314**
See Bug 107963, and ZCS-1461

### CODE REVIEW

File: conf/nginx/nginx.conf.web.https.default.template [hunk 1]
```diff
@@ -1,12 +1,33 @@
...
-    ${core.ipboth.enabled}listen                  [::]:${web.https.port} default ipv6only=off;
-    ${core.ipv4only.enabled}listen                ${web.https.port} default;
-    ${core.ipv6only.enabled}listen                [::]:${web.https.port} default;
+    ${core.ipboth.enabled}listen                  [::]:${web.https.port} ipv6only=off;
+    ${core.ipv4only.enabled}listen                ${web.https.port};
+    ${core.ipv6only.enabled}listen                [::]:${web.https.port};
+
     ${web.add.headers.default}
-    server_name             ${web.server_name.default}.default;
+    server_name             ${web.server_name.default}; # add aliases and perhaps public
```
Comment (shri314) [inline conf/nginx/nginx.conf.web.https.default.template:30]: I don't know if there is any significance to removing the .default from the server_name. I have tested with and without it, and it seems to be working just fine. From perforce, it seems like we had the extra .default way back when this config was created. But it seems like a typo.

File: conf/nginx/nginx.conf.web.https.default.template [hunk 2]
```diff
@@ -1,12 +1,33 @@
 # HTTPS Proxy Default Configuration
 #
+server {
+    ${core.ipboth.enabled}listen                  [::]:${web.https.port} default_server ipv6only=off;
```
Comment (shri314) [inline conf/nginx/nginx.conf.web.https.default.template:4]: Marking this "server" as the "default_server"

File: conf/nginx/nginx.conf.web.https.default.template [hunk 3]
```diff
@@ -1,12 +1,33 @@
...
+    ssl_session_cache       ${ssl.session.cachesize};
+    ssl_session_timeout     ${ssl.session.timeout};
+    ssl_ciphers             ${web.ssl.ciphers};
+    ssl_ecdh_curve          ${web.ssl.ecdh.curve};
+    ssl_certificate         ${ssl.crt.default};
+    ssl_certificate_key     ${ssl.key.default};
+    ssl_verify_client       ${ssl.clientcertmode.default};
+    ssl_verify_depth        ${ssl.clientcertdepth.default};
+    ${web.ssl.dhparam.enabled}ssl_dhparam             ${web.ssl.dhparam.file};
+    return 444;
```
Comment (shri314) [inline conf/nginx/nginx.conf.web.https.default.template:20]: This is the main part of the fix. The default_server will return 444.

File: conf/nginx/nginx.conf.web.https.default.template [hunk 4]
```diff
@@ -1,12 +1,33 @@
...
+}
+
 server
 {
-    ${core.ipboth.enabled}listen                  [::]:${web.https.port} default ipv6only=off;
-    ${core.ipv4only.enabled}listen                ${web.https.port} default;
-    ${core.ipv6only.enabled}listen                [::]:${web.https.port} default;
+    ${core.ipboth.enabled}listen                  [::]:${web.https.port} ipv6only=off;
```
Comment (shri314) [inline conf/nginx/nginx.conf.web.https.default.template:25]: Removed this "server" as the "default".

**Comment by shri314** [comment]
Tests:
1) Tested default installation, log in.
2) Set up virtual servers using virtual interface, log in with virtual server domain.
3) Set up virtual servers using SNI, log in with virtual server domain.

In all cases vulnerability was checked from command line. With the fix - The server response is empty. Without the fix - the server responded with "HTTP/1.1 302 Found" with the spoofed "Location" header.

For details about SNI -
 refer to https://wiki.zimbra.com/wiki/Multiple_SSL_Certificates,_Server_Name_Indication_(SNI)_for_HTTPS

**Comment by Prashantsurana** [comment]
@plobbes @gordyt  awaiting your review before committing the changes?

**Review by gordyt — APPROVED** [review]
Looks good!

**Review by tmclane — APPROVED** [review]

File: conf/nginx/nginx.conf.web.https.default.template (hunk 1 above)
Comment (plobbes) [reply conf/nginx/nginx.conf.web.https.default.template:30]: Agreed. I didn't see any obvious reason to keep it.  In fact, removing the .default seemed the appropriate change to have server_name set to a valid FQDN.

**Review by plobbes — APPROVED** [review]
LGTM.

**Comment by ozzi-** [comment]
Is this fixed on a 8.7.11 upgrade?

**Comment by ozzi-** [comment]
Shouldn't those changes also be made to nginx.conf.web.http.default.template instead of only the https template?

**Pull request merged**